echo-notifications/src/main/groovy/com/netflix/spinnaker/echo/notification/InteractiveNotificationCallbackHandler.groovy

```diff
@@ -0,0 +1,97 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.echo.notification
+
+import com.jakewharton.retrofit.Ok3Client
+import com.netflix.spinnaker.echo.api.Notification
+import com.netflix.spinnaker.echo.api.Notification.InteractiveActionCallback
+import com.netflix.spinnaker.retrofit.Slf4jRetrofitLogger
+import groovy.util.logging.Slf4j
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.core.env.Environment
+import org.springframework.http.HttpHeaders
+import org.springframework.stereotype.Component
+import retrofit.RestAdapter
+import retrofit.client.Response
+import retrofit.converter.JacksonConverter
+import retrofit.http.Body
+import retrofit.http.Header
+import retrofit.http.POST
+
+import static retrofit.Endpoints.newFixedEndpoint
+
+@Slf4j
+@Component
+class InteractiveNotificationCallbackHandler {
+  @Autowired
+  Ok3Client spinnakerServiceClient
+
+  @Autowired
+  List<InteractiveNotificationService> notificationServices
+
+  @Autowired
+  private Environment environment
+
+  /**
+   * Processes a callback request from Slack by relaying it to the downstream service which originated the message
+   * referenced in the payload.
+   * @param event
+   */
+  Map processCallback(String source, Map content, HttpHeaders headers) {
+    log.debug("Received interactive notification callback request from ${source}")
+
+    InteractiveNotificationService notificationService = notificationServices.stream()
+      .find { it.supportsType(Notification.Type.valueOf(source.toUpperCase())) }
+
+    if (notificationService == null) {
+      throw new IllegalArgumentException("NotificationService for ${source} no registered")
+    }
+
+    InteractiveActionCallback callback = notificationService.parseInteractionCallback(content)
+    String requestUrl = environment.getProperty("${callback.serviceId}.baseUrl") + "/notifications/callback"
+    SpinnakerService spinnakerService = new RestAdapter.Builder()
+      .setEndpoint(newFixedEndpoint(requestUrl))
+      .setClient(spinnakerServiceClient)
+      .setLogLevel(RestAdapter.LogLevel.BASIC)
+      .setLog(new Slf4jRetrofitLogger(SpinnakerService.class))
+      .setConverter(new JacksonConverter())
+      .build()
+      .create(SpinnakerService.class)
+
+    log.debug("Routing notification callback to originating service at ${requestUrl}")
+
+    // TODO: error handling (retries?)
+    //    retrySupport.retry(
+    //      () ->
+    Response response = spinnakerService.post(callback, callback.user)
+    log.debug("Received callback response from downstream Spinnaker service: ${response}")
+    //      5,
+    //      2000,
+    //      false);
+    //  }
+
+    // Allows the notification service implementation to respond to the callback as needed
+    notificationService.respondToCallback(content)
+
+    return null
```

Comment:
For Slack, this is just an acknowledgement so there's no body to return.

Reply:
Hmm, what would be in this map if the function were to return a map?  Maybe the return type should be void for now?  I think returning `null` due to the one Slack implementation should probably be avoided.

Reply:
I have no idea what the other types of services we integrate with expect from these kinds of webhook interactions, so I was trying to make it as generic as possible -- take a JSON payload, return a JSON payload. If you got nothing to return, return null. Could change this to an `Optional` perhaps? Do RestController + jackson handle that automagically like they do nulls?

Reply:
I'm not sure if there's any automagic stuff regarding `RestController` and `Optional`, but let's use `Optional` here so we don't have to refactor this later when there is another interactive notification implementation that returns something.